### ISSUE TITLE
Improper handling of partial type stub packages

### ISSUE THREAD

**Issue by erictraut** [microsoft/pylance-release#5139]
Pyright isn't properly handling type stub packages (i.e. a package that is named "x-stubs") if the "py.typed" file indicates that it's "partial". Import resolution is failing due to an apparent bug in the readonlyAugmentedFileSystem that was added to support the shadowing of implementation source files.

To repro, `pip install types-tensorflow`. Then create a new local source file with the following contents:
```python
import tensorflow as tf

reveal_type(tf)
reveal_type(tf.keras)
```

The `tf.keras` should be evaluated as a module, but it's not. This is the first symptom of this bug.

Open the `site-packages/tensorflow-stubs/__init__.pyi` file in the editor. Note that there is an error reported on the line `from tensorflow.keras import losses as losses`. This should not generate an error and should resolve correctly. This is a second symptom of this bug.

Next, open the file `site-packages/tensorflow-stubs/keras/__init__.pyi` in the editor. Note that there is a similar import resolution error reported here. This is a third symptom of this bug.

Comment out the multiline `from tensorflow.keras import { ...` statement. Above it, type `from . import layers as layers`. Go to the output panel for pyright or pylance to look at the log output. Note that it contains the log: `Received change text document command for closed file /<redacted>/site-packages/tensorflow/keras/__init__.pyi`. This is a fourth symptom of this bug.

I don't fully understand the readonlyAugmentedFileSystem and its use of "moved entries", but it looks like the problem is related to this mechanism. 

@heejaechang, I think you wrote this code, so you're probably the most qualified to diagnose and fix this issue.

This is blocking an [issue in typeshed](https://github.com/python/typeshed/issues/11052), so I'd appreciate it if you'd make this a priority.

**Comment by rchiodo** [comment]
This looks to be a problem in our moved entries not mapping directories. So the `keras` directory is not mapped and we don't find it when looking for imports.

**Comment by heejaechang** [comment]
it is fixed in pyright, didn't pulled into pylance yet.

**Comment by debonte** [comment]
This issue has been fixed in prerelease version 2023.12.100, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202312100-14-December-2023-PreRelease)
